File: go/crypto.go
package samp

import (
	"crypto/rand"
	"crypto/sha256"
	"crypto/sha512"
	"errors"
	"io"

	"github.com/gtank/ristretto255"
	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/hkdf"
)

const EncryptedOverhead = 80

var (
	ErrDecryptionFailed = errors.New("samp: decryption failed")
	ErrInvalidPoint     = errors.New("samp: invalid ristretto255 point")
)

// WHY: the single crypto boundary that turns a 32-byte ViewScalar into a
// ristretto255.Scalar. Every decrypt path funnels through here.
func viewScalarToRistretto(v ViewScalar) *ristretto255.Scalar {
	raw := v.b
	s := ristretto255.NewScalar()
	if err := s.Decode(raw[:]); err == nil {
		return s
	}
	var wide [64]byte
	copy(wide[:32], raw[:])
	return new(ristretto255.Scalar).FromUniformBytes(wide[:])
}

func Sr25519SigningScalar(seed Seed) ViewScalar {
	raw := seed.b
	h := sha512.Sum512(raw[:])
	h[0] &= 248
	h[31] &= 63
	h[31] |= 64
	divideScalarByCofactor(h[:32])
	s := ristretto255.NewScalar()
	if err := s.Decode(h[:32]); err != nil {
		var wide [64]byte
		copy(wide[:32], h[:32])
		s = new(ristretto255.Scalar).FromUniformBytes(wide[:])
	}
	var out [32]byte
	copy(out[:], s.Encode(nil))
	return ViewScalar{out}
}

func divideScalarByCofactor(s []byte) {
	var low byte
	for i := len(s) - 1; i >= 0; i-- {
		r := s[i] & 0x07
		s[i] >>= 3
		s[i] += low
		low = r << 5
	}
}

func PublicFromSeed(seed Seed) Pubkey {
	vs := Sr25519SigningScalar(seed)
	pub := new(ristretto255.Element).ScalarBaseMult(viewScalarToRistretto(vs))
	var out [32]byte
	copy(out[:], pub.Encode(nil))
	return Pubkey{out}
}

func hkdfExpand(ikm, salt, info []byte, length int) []byte {
	r := hkdf.New(sha256.New, ikm, salt, info)
	out := make([]byte, length)
	io.ReadFull(r, out)
	return out
}

func deriveEphemeral(seed Seed, recipient [32]byte, nonce Nonce) []byte {
	info := make([]byte, 44)
	copy(info[:32], recipient[:])
	copy(info[32:], nonce.b[:])
	raw := seed.b
	return hkdfExpand(raw[:], nil, info, 32)
}

func deriveSealKey(seed Seed, nonce Nonce) []byte {
	raw := seed.b
	return hkdfExpand(raw[:], nonce.chachaNonce(), []byte("samp-seal"), 32)
}

func deriveSymmetricKey(sharedSecret []byte, nonce Nonce) []byte {
	return hkdfExpand(sharedSecret, nonce.chachaNonce(), []byte("samp-message"), 32)
}

func deriveViewTagByte(sharedSecret []byte) byte {
	out := hkdfExpand(sharedSecret, nil, []byte("samp-view-tag"), 1)
	return out[0]
}

func ecdhSharedSecret(scalar *ristretto255.Scalar, pointBytes []byte) ([]byte, error) {
	p := new(ristretto255.Element)
	if err := p.Decode(pointBytes); err != nil {
		return nil, ErrInvalidPoint
	}
	shared := new(ristretto255.Element).ScalarMult(scalar, p)
	return shared.Encode(nil), nil
}

func scalarFromBytes(b []byte) *ristretto255.Scalar {
	var wide [64]byte
	copy(wide[:32], b)
	return new(ristretto255.Scalar).FromUniformBytes(wide[:])
}

func Encrypt(plaintext Plaintext, recipientPub Pubkey, nonce Nonce, senderSeed Seed) (Ciphertext, error) {
	recBytes := recipientPub.b
	ephBytes := deriveEphemeral(senderSeed, recBytes, nonce)
	ephScalar := scalarFromBytes(ephBytes)
	ephPubkey := new(ristretto255.Element).ScalarBaseMult(ephScalar)

	sharedSecret, err := ecdhSharedSecret(ephScalar, recBytes[:])
	if err != nil {
		return Ciphertext{}, err
	}

	sealKey := deriveSealKey(senderSeed, nonce)
	var sealedTo [32]byte
	for i := 0; i < 32; i++ {
		sealedTo[i] = recBytes[i] ^ sealKey[i]
	}

	symKey := deriveSymmetricKey(sharedSecret, nonce)
	aead, err := chacha20poly1305.New(symKey)
	if err != nil {
		return Ciphertext{}, err
	}
	ciphertextWithTag := aead.Seal(nil, nonce.chachaNonce(), plaintext.b, sealedTo[:])

	out := make([]byte, 0, EncryptedOverhead+len(plaintext.b))
	out = append(out, ephPubkey.Encode(nil)...)
	out = append(out, sealedTo[:]...)
	out = append(out, ciphertextWithTag...)
	return Ciphertext{out}, nil
}

func Decrypt(ct Ciphertext, nonce Nonce, signingScalar ViewScalar) (Plaintext, error) {
	if len(ct.b) < EncryptedOverhead {
		return Plaintext{}, ErrInsufficientData
	}
	content := ct.b
	sharedSecret, err := ecdhSharedSecret(viewScalarToRistretto(signingScalar), content[:32])
	if err != nil {
		return Plaintext{}, err
	}
	symKey := deriveSymmetricKey(sharedSecret, nonce)
	aead, err := chacha20poly1305.New(symKey)
	if err != nil {
		return Plaintext{}, err
	}
	plaintext, err := aead.Open(nil, nonce.chachaNonce(), content[64:], content[32:64])
	if err != nil {
		return Plaintext{}, ErrDecryptionFailed
	}
	return Plaintext{plaintext}, nil
}

func DecryptAsSender(ct Ciphertext, nonce Nonce, senderSeed Seed) (Plaintext, error) {
	if len(ct.b) < EncryptedOverhead {
		return Plaintext{}, ErrInsufficientData
	}
	content := ct.b
	sealKey := deriveSealKey(senderSeed, nonce)
	var recipient [32]byte
	for i := 0; i < 32; i++ {
		recipient[i] = content[32+i] ^ sealKey[i]
	}

	ephBytes := deriveEphemeral(senderSeed, recipient, nonce)
	ephScalar := scalarFromBytes(ephBytes)
	sharedSecret, err := ecdhSharedSecret(ephScalar, recipient[:])
	if err != nil {
		return Plaintext{}, err
	}

	symKey := deriveSymmetricKey(sharedSecret, nonce)
	aead, err := chacha20poly1305.New(symKey)
	if err != nil {
		return Plaintext{}, err
	}
	plaintext, err := aead.Open(nil, nonce.chachaNonce(), content[64:], content[32:64])
	if err != nil {
		return Plaintext{}, ErrDecryptionFailed
	}
	return Plaintext{plaintext}, nil
}

func CheckViewTag(ct Ciphertext, signingScalar ViewScalar) (ViewTag, error) {
	if len(ct.b) < EncryptedOverhead {
		return ViewTag{}, ErrInsufficientData
	}
	sharedSecret, err := ecdhSharedSecret(viewScalarToRistretto(signingScalar), ct.b[:32])
	if err != nil {
		return ViewTag{}, err
	}
	return ViewTag{deriveViewTagByte(sharedSecret)}, nil
}

func UnsealRecipient(ct Ciphertext, nonce Nonce, senderSeed Seed) (Pubkey, error) {
	if len(ct.b) < EncryptedOverhead {
		return Pubkey{}, ErrInsufficientData
	}
	sealKey := deriveSealKey(senderSeed, nonce)
	var recipient [32]byte
	for i := 0; i < 32; i++ {
		recipient[i] = ct.b[32+i] ^ sealKey[i]
	}
	return Pubkey{recipient}, nil
}

func ComputeViewTag(senderSeed Seed, recipient Pubkey, nonce Nonce) (ViewTag, error) {
	recBytes := recipient.b
	ephBytes := deriveEphemeral(senderSeed, recBytes, nonce)
	ephScalar := scalarFromBytes(ephBytes)
	sharedSecret, err := ecdhSharedSecret(ephScalar, recBytes[:])
	if err != nil {
		return ViewTag{}, err
	}
	return ViewTag{deriveViewTagByte(sharedSecret)}, nil
}

func deriveGroupEphemeral(senderSeed Seed, nonce Nonce) []byte {
	info := make([]byte, 0, len("samp-group-eph")+12)
	info = append(info, []byte("samp-group-eph")...)
	info = append(info, nonce.b[:]...)
	raw := senderSeed.b
	return hkdfExpand(raw[:], nil, info, 32)
}

func deriveKeyWrap(sharedSecret []byte, nonce Nonce) []byte {
	return hkdfExpand(sharedSecret, nonce.chachaNonce(), []byte("samp-key-wrap"), 32)
}

func buildCapsules(contentKey ContentKey, members []Pubkey, ephScalar *ristretto255.Scalar, nonce Nonce) Capsules {
	out := make([]byte, 0, len(members)*CapsuleSize)
	ck := contentKey.b
	for _, pk := range members {
		pkb := pk.b
		shared, err := ecdhSharedSecret(ephScalar, pkb[:])
		if err != nil {
			out = append(out, make([]byte, CapsuleSize)...)
			continue
		}
		tag := deriveViewTagByte(shared)
		kek := deriveKeyWrap(shared, nonce)
		out = append(out, tag)
		for i := 0; i < 32; i++ {
			out = append(out, ck[i]^kek[i])
		}
	}
	return Capsules{out}
}

func scanCapsules(data []byte, ephPubkey EphPubkey, myScalar ViewScalar, nonce Nonce) (int, ContentKey, bool) {
	epb := ephPubkey.b
	shared, err := ecdhSharedSecret(viewScalarToRistretto(myScalar), epb[:])
	if err != nil {
		return 0, ContentKey{}, false
	}
	myTag := deriveViewTagByte(shared)
	kek := deriveKeyWrap(shared, nonce)

	offset := 0
	idx := 0
	for offset+CapsuleSize <= len(data) {
		if data[offset] == myTag {
			var ck [32]byte
			for i := 0; i < 32; i++ {
				ck[i] = data[offset+1+i] ^ kek[i]
			}
			return idx, ContentKey{ck}, true
		}
		offset += CapsuleSize
		idx++
	}
	return 0, ContentKey{}, false
}

func EncryptForGroup(plaintext Plaintext, members []Pubkey, nonce Nonce, senderSeed Seed) (EphPubkey, Capsules, Ciphertext, error) {
	ephBytes := deriveGroupEphemeral(senderSeed, nonce)
	ephScalar := scalarFromBytes(ephBytes)
	ephPub := new(ristretto255.Element).ScalarBaseMult(ephScalar)

	var ck [32]byte
	if _, err := rand.Read(ck[:]); err != nil {
		return EphPubkey{}, Capsules{}, Ciphertext{}, err
	}
	contentKey := ContentKey{ck}

	capsules := buildCapsules(contentKey, members, ephScalar, nonce)

	aead, err := chacha20poly1305.New(ck[:])
	if err != nil {
		return EphPubkey{}, Capsules{}, Ciphertext{}, err
	}
	ct := aead.Seal(nil, nonce.chachaNonce(), plaintext.b, nil)

	var ephArr [32]byte
	copy(ephArr[:], ephPub.Encode(nil))
	return EphPubkey{ephArr}, capsules, Ciphertext{ct}, nil
}

func DecryptFromGroup(content []byte, myScalar ViewScalar, nonce Nonce, knownN int) (Plaintext, error) {
	if len(content) < 32 {
		return Plaintext{}, ErrInsufficientData
	}
	var ephArr [32]byte
	copy(ephArr[:], content[:32])
	ephPubkey := EphPubkey{ephArr}
	afterEph := content[32:]

	capsuleIdx, contentKey, found := scanCapsules(afterEph, ephPubkey, myScalar, nonce)
	if !found {
		return Plaintext{}, ErrDecryptionFailed
	}
	ckRaw := contentKey.b

	aead, err := chacha20poly1305.New(ckRaw[:])
	if err != nil {
		return Plaintext{}, err
	}

	if knownN > 0 {
		ctStart := knownN * CapsuleSize
		if ctStart > len(afterEph) {
			return Plaintext{}, ErrInsufficientData
		}
		pt, err := aead.Open(nil, nonce.chachaNonce(), afterEph[ctStart:], nil)
		if err != nil {
			return Plaintext{}, ErrDecryptionFailed
		}
		return Plaintext{pt}, nil
	}

	minN := capsuleIdx + 1
	maxN := (len(afterEph) - 16) / CapsuleSize
	for n := minN; n <= maxN; n++ {
		ctStart := n * CapsuleSize
		if ctStart >= len(afterEph) {
			break
		}
		pt, err := aead.Open(nil, nonce.chachaNonce(), afterEph[ctStart:], nil)
		if err == nil {
			return Plaintext{pt}, nil
		}
	}
	return Plaintext{}, ErrDecryptionFailed
}

File: typescript/src/crypto.ts
import { sha512 } from "@noble/hashes/sha512";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";
import { chacha20poly1305 } from "@noble/ciphers/chacha";
import { RistrettoPoint } from "@noble/curves/ed25519";
import { bytesToNumberLE, numberToBytesLE } from "@noble/curves/abstract/utils";
import { SampError } from "./error.js";
import { ContentKey, Seed, ViewScalar } from "./secret.js";
import { Capsules, Ciphertext, EphPubkey, Nonce, Plaintext, Pubkey, Signature, ViewTag } from "./types.js";
import { sr25519KeypairFromSeed, sr25519Sign as wasmSr25519Sign } from "@polkadot/wasm-crypto";

export const ENCRYPTED_OVERHEAD = 80;
const CAPSULE_SIZE = 33;
const CURVE_ORDER = BigInt(
  "7237005577332262213973186563042994240857116359379907606001950938285454250989",
);

const MESSAGE_KEY_INFO = new TextEncoder().encode("samp-message");
const VIEW_TAG_INFO = new TextEncoder().encode("samp-view-tag");
const SEAL_INFO = new TextEncoder().encode("samp-seal");
const GROUP_EPH_INFO = new TextEncoder().encode("samp-group-eph");
const KEY_WRAP_INFO = new TextEncoder().encode("samp-key-wrap");

function mod(n: bigint, m: bigint): bigint {
  return ((n % m) + m) % m;
}

function divideScalarByCofactor(s: Uint8Array): void {
  let low = 0;
  for (let i = s.length - 1; i >= 0; i--) {
    const r = s[i]! & 0x07;
    s[i] = (s[i]! >>> 3) + low;
    low = r << 5;
  }
}

function hkdfExpand(ikm: Uint8Array, salt: Uint8Array | undefined, info: Uint8Array, length: number): Uint8Array {
  return hkdf(sha256, ikm, salt, info, length);
}

function scalarFromBytes(b: Uint8Array): bigint {
  return mod(bytesToNumberLE(b), CURVE_ORDER);
}

function ecdhSharedSecret(scalar: bigint, pointBytes: Uint8Array): Uint8Array {
  return RistrettoPoint.fromHex(pointBytes).multiply(scalar).toRawBytes();
}

function xor(a: Uint8Array, b: Uint8Array): Uint8Array {
  const out = new Uint8Array(a.length);
  for (let i = 0; i < a.length; i++) out[i] = a[i]! ^ b[i]!;
  return out;
}

function deriveEphemeral(seed: Seed, recipient: Pubkey, nonce: Nonce): Uint8Array {
  const info = new Uint8Array(44);
  info.set(recipient, 0);
  info.set(Nonce.chachaBytes(nonce), 32);
  return hkdfExpand(Seed.exposeSecret(seed), undefined, info, 32);
}

function deriveSealKey(seed: Seed, nonce: Nonce): Uint8Array {
  return hkdfExpand(Seed.exposeSecret(seed), Nonce.chachaBytes(nonce), SEAL_INFO, 32);
}

function deriveSymmetricKey(sharedSecret: Uint8Array, nonce: Nonce): Uint8Array {
  return hkdfExpand(sharedSecret, Nonce.chachaBytes(nonce), MESSAGE_KEY_INFO, 32);
}

function deriveViewTagByte(sharedSecret: Uint8Array): number {
  return hkdfExpand(sharedSecret, undefined, VIEW_TAG_INFO, 1)[0]!;
}

function deriveKeyWrap(sharedSecret: Uint8Array, nonce: Nonce): Uint8Array {
  return hkdfExpand(sharedSecret, Nonce.chachaBytes(nonce), KEY_WRAP_INFO, 32);
}

// WHY: the single crypto boundary that turns a 32-byte ViewScalar into the
// bigint scalar the ristretto curve operations expect.
function viewScalarToBigInt(vs: ViewScalar): bigint {
  return mod(bytesToNumberLE(ViewScalar.exposeSecret(vs)), CURVE_ORDER);
}

export function sr25519Sign(seed: Seed, message: Uint8Array): Signature {
  const kp = sr25519KeypairFromSeed(Seed.exposeSecret(seed));
  const sig = wasmSr25519Sign(kp.slice(64, 96), kp.slice(0, 64), message);
  return Signature.fromBytes(sig);
}

export function sr25519SigningScalar(seed: Seed): ViewScalar {
  const raw = Seed.exposeSecret(seed);
  const h = sha512(raw);
  h[0]! &= 248;
  h[31]! &= 63;
  h[31]! |= 64;
  const key = h.slice(0, 32);
  divideScalarByCofactor(key);
  const scalar = mod(bytesToNumberLE(key), CURVE_ORDER);
  return ViewScalar.fromBytes(numberToBytesLE(scalar, 32));
}

export function publicFromSeed(seed: Seed): Pubkey {
  const scalar = sr25519SigningScalar(seed);
  const raw = RistrettoPoint.BASE.multiply(viewScalarToBigInt(scalar)).toRawBytes();
  return Pubkey.fromBytes(raw);
}

export function encrypt(
  plaintext: Plaintext,
  recipient: Pubkey,
  nonce: Nonce,
  senderSeed: Seed,
): Ciphertext {
  const ephBytes = deriveEphemeral(senderSeed, recipient, nonce);
  const ephScalar = scalarFromBytes(ephBytes);
  const ephPubkey = RistrettoPoint.BASE.multiply(ephScalar).toRawBytes();
  const sharedSecret = ecdhSharedSecret(ephScalar, recipient);
  const sealKey = deriveSealKey(senderSeed, nonce);
  const sealedTo = xor(recipient, sealKey);
  const symKey = deriveSymmetricKey(sharedSecret, nonce);
  const cipher = chacha20poly1305(symKey, Nonce.chachaBytes(nonce), sealedTo);
  const ptBytes = Plaintext.asBytes(plaintext);
  const ctWithTag = cipher.encrypt(ptBytes);
  const out = new Uint8Array(ENCRYPTED_OVERHEAD + ptBytes.length);
  out.set(ephPubkey, 0);
  out.set(sealedTo, 32);
  out.set(ctWithTag, 64);
  return Ciphertext.fromBytes(out);
}

export function decrypt(ciphertext: Ciphertext, nonce: Nonce, signingScalar: ViewScalar): Plaintext {
  const content = Ciphertext.asBytes(ciphertext);
  if (content.length < ENCRYPTED_OVERHEAD) throw new SampError("insufficient data");
  const sharedSecret = ecdhSharedSecret(viewScalarToBigInt(signingScalar), content.slice(0, 32));
  const sealedTo = content.slice(32, 64);
  const symKey = deriveSymmetricKey(sharedSecret, nonce);
  const cipher = chacha20poly1305(symKey, Nonce.chachaBytes(nonce), sealedTo);
  try {
    return Plaintext.fromBytes(cipher.decrypt(content.slice(64)));
  } catch {
    throw new SampError("decryption failed");
  }
}

export function decryptAsSender(ciphertext: Ciphertext, nonce: Nonce, senderSeed: Seed): Plaintext {
  const content = Ciphertext.asBytes(ciphertext);
  if (content.length < ENCRYPTED_OVERHEAD) throw new SampError("insufficient data");
  const sealKey = deriveSealKey(senderSeed, nonce);
  const recipientBytes = xor(content.slice(32, 64), sealKey);
  const recipient = Pubkey.fromBytes(recipientBytes);
  const ephBytes = deriveEphemeral(senderSeed, recipient, nonce);
  const ephScalar = scalarFromBytes(ephBytes);
  const sharedSecret = ecdhSharedSecret(ephScalar, recipient);
  const symKey = deriveSymmetricKey(sharedSecret, nonce);
  const sealedTo = content.slice(32, 64);
  const cipher = chacha20poly1305(symKey, Nonce.chachaBytes(nonce), sealedTo);
  try {
    return Plaintext.fromBytes(cipher.decrypt(content.slice(64)));
  } catch {
    throw new SampError("decryption failed");
  }
}

export function computeViewTag(senderSeed: Seed, recipient: Pubkey, nonce: Nonce): ViewTag {
  const ephBytes = deriveEphemeral(senderSeed, recipient, nonce);
  const ephScalar = scalarFromBytes(ephBytes);
  const sharedSecret = ecdhSharedSecret(ephScalar, recipient);
  return ViewTag.from(deriveViewTagByte(sharedSecret));
}

export function checkViewTag(ciphertext: Ciphertext, signingScalar: ViewScalar): ViewTag {
  const content = Ciphertext.asBytes(ciphertext);
  if (content.length < ENCRYPTED_OVERHEAD) throw new SampError("insufficient data");
  const sharedSecret = ecdhSharedSecret(viewScalarToBigInt(signingScalar), content.slice(0, 32));
  return ViewTag.from(deriveViewTagByte(sharedSecret));
}

export function unsealRecipient(ciphertext: Ciphertext, nonce: Nonce, senderSeed: Seed): Pubkey {
  const content = Ciphertext.asBytes(ciphertext);
  if (content.length < ENCRYPTED_OVERHEAD) throw new SampError("insufficient data");
  const sealKey = deriveSealKey(senderSeed, nonce);
  return Pubkey.fromBytes(xor(content.slice(32, 64), sealKey));
}

function deriveGroupEphemeralBytes(senderSeed: Seed, nonce: Nonce): Uint8Array {
  const info = new Uint8Array(GROUP_EPH_INFO.length + 12);
  info.set(GROUP_EPH_INFO, 0);
  info.set(Nonce.chachaBytes(nonce), GROUP_EPH_INFO.length);
  return hkdfExpand(Seed.exposeSecret(senderSeed), undefined, info, 32);
}

export function deriveGroupEphemeral(senderSeed: Seed, nonce: Nonce): Uint8Array {
  const bytes = deriveGroupEphemeralBytes(senderSeed, nonce);
  return numberToBytesLE(scalarFromBytes(bytes), 32);
}

export function buildCapsules(
  contentKey: ContentKey,
  memberPubkeys: Pubkey[],
  ephScalarBytes: Uint8Array,
  nonce: Nonce,
): Capsules {
  const scalar = scalarFromBytes(ephScalarBytes);
  const ck = ContentKey.exposeSecret(contentKey);
  const out = new Uint8Array(memberPubkeys.length * CAPSULE_SIZE);
  for (let i = 0; i < memberPubkeys.length; i++) {
    const shared = ecdhSharedSecret(scalar, memberPubkeys[i]!);
    const tag = deriveViewTagByte(shared);
    const kek = deriveKeyWrap(shared, nonce);
    out[i * CAPSULE_SIZE] = tag;
    for (let j = 0; j < 32; j++) {
      out[i * CAPSULE_SIZE + 1 + j] = ck[j]! ^ kek[j]!;
    }
  }
  return Capsules.fromBytes(out);
}

function scanCapsules(
  data: Uint8Array,
  ephPubkey: EphPubkey,
  myScalar: ViewScalar,
  nonce: Nonce,
): { index: number; contentKey: ContentKey } | null {
  const shared = ecdhSharedSecret(viewScalarToBigInt(myScalar), ephPubkey);
  const myTag = deriveViewTagByte(shared);
  const kek = deriveKeyWrap(shared, nonce);
  let offset = 0;
  let idx = 0;
  while (offset + CAPSULE_SIZE <= data.length) {
    if (data[offset] === myTag) {
      const out = new Uint8Array(32);
      for (let j = 0; j < 32; j++) out[j] = data[offset + 1 + j]! ^ kek[j]!;
      return { index: idx, contentKey: ContentKey.fromBytes(out) };
    }
    offset += CAPSULE_SIZE;
    idx++;
  }
  return null;
}

export function encryptForGroup(
  plaintext: Plaintext,
  memberPubkeys: Pubkey[],
  nonce: Nonce,
  senderSeed: Seed,
): { ephPubkey: EphPubkey; capsules: Capsules; ciphertext: Ciphertext } {
  const ephBytes = deriveGroupEphemeralBytes(senderSeed, nonce);
  const ephScalar = scalarFromBytes(ephBytes);
  const ephPubkey = EphPubkey.fromBytes(RistrettoPoint.BASE.multiply(ephScalar).toRawBytes());
  const ckRaw = crypto.getRandomValues(new Uint8Array(32));
  const contentKey = ContentKey.fromBytes(ckRaw);
  const capsules = buildCapsules(contentKey, memberPubkeys, ephBytes, nonce);
  const cipher = chacha20poly1305(ckRaw, Nonce.chachaBytes(nonce));
  const ct = cipher.encrypt(Plaintext.asBytes(plaintext));
  return { ephPubkey, capsules, ciphertext: Ciphertext.fromBytes(ct) };
}

export function decryptFromGroup(
  content: Uint8Array,
  myScalar: ViewScalar,
  nonce: Nonce,
  knownN?: number,
): Plaintext {
  if (content.length < 32) throw new SampError("insufficient data");
  const ephPubkey = EphPubkey.fromBytes(content.slice(0, 32));
  const afterEph = content.slice(32);
  const scan = scanCapsules(afterEph, ephPubkey, myScalar, nonce);
  if (scan === null) throw new SampError("decryption failed");
  const ckRaw = ContentKey.exposeSecret(scan.contentKey);
  const cipher = chacha20poly1305(ckRaw, Nonce.chachaBytes(nonce));
  if (knownN !== undefined) {
    const ctStart = knownN * CAPSULE_SIZE;
    if (ctStart > afterEph.length) throw new SampError("insufficient data");
    try {
      return Plaintext.fromBytes(cipher.decrypt(afterEph.slice(ctStart)));
    } catch {
      throw new SampError("decryption failed");
    }
  }
  const minN = scan.index + 1;
  const maxN = Math.floor((afterEph.length - 16) / CAPSULE_SIZE);
  for (let n = minN; n <= maxN; n++) {
    const ctStart = n * CAPSULE_SIZE;
    if (ctStart >= afterEph.length) break;
    const c = chacha20poly1305(ckRaw, Nonce.chachaBytes(nonce));
    try {
      return Plaintext.fromBytes(c.decrypt(afterEph.slice(ctStart)));
    } catch {
      continue;
    }
  }
  throw new SampError("decryption failed");
}
